### PR TITLE
SONARJAVA-2575 RSPEC-4348 Iterable iterator returning this

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/IterableIteratorCheck.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/IterableIteratorCheck.java]: MINOR Remove this unused import 'org.sonar.plugins.java.api.tree.BinaryExpressionTree'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: java-checks/src/main/java/org/sonar/java/checks/IterableIteratorCheck.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/IterableIteratorCheck.java]: MINOR Remove this unused import 'org.sonar.plugins.java.api.tree.ExpressionTree'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: java-checks/src/main/java/org/sonar/java/checks/IterableIteratorCheck.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/IterableIteratorCheck.java]: MINOR Remove this unused import 'com.google.common.collect.ImmutableMap'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: java-checks/src/main/java/org/sonar/java/checks/IterableIteratorCheck.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.ReturnStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.UnaryExpressionTree;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/IterableIteratorCheck.java]: MINOR Remove this unused import 'org.sonar.plugins.java.api.tree.UnaryExpressionTree'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: java-checks/src/main/java/org/sonar/java/checks/IterableIteratorCheck.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/IterableIteratorCheck.java]: MINOR Remove this unused import 'java.util.Map'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: java-checks/src/main/java/org/sonar/java/checks/IterableIteratorCheck.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.java.model.ExpressionUtils;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/IterableIteratorCheck.java]: MINOR Remove this unused import 'org.sonar.java.model.ExpressionUtils'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* ![MINOR][MINOR] 6 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: java-checks/src/main/java/org/sonar/java/checks/IterableIteratorCheck.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    if (!hasSemantic()) {
+      return;
+    }
+    ReturnThis returnThis = new ReturnThis();
+    ClassTree classTree = (ClassTree) tree;
+    Type classType = classTree.symbol().type();
+    if(!(classType.isSubtypeOf("java.util.Iterator") && classType.isSubtypeOf("java.lang.Iterable"))) {
+      return;
+    }
+    classTree.members().stream().filter(m -> m.is(Tree.Kind.METHOD)).forEach(t -> t.accept(returnThis));
```
Comment (saberduck) [inline java-checks/src/main/java/org/sonar/java/checks/IterableIteratorCheck.java]: if I am not mistaken, it should only check for `iterator()` method, not all methods. At least RSPEC is worded that way

File: java-checks/src/main/java/org/sonar/java/checks/IterableIteratorCheck.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    Type classType = classTree.symbol().type();
+    if(!(classType.isSubtypeOf("java.util.Iterator") && classType.isSubtypeOf("java.lang.Iterable"))) {
+      return;
+    }
+    classTree.members().stream()
+      .filter(m -> m.is(Tree.Kind.METHOD))
+      .filter(m -> ITERATOR.matches(((MethodTree) m)))
+      .forEach(t -> t.accept(returnThis));
+    if (!returnThis.issueLocations.isEmpty()) {
+      reportIssue(returnThis.issueLocations.get(0), "",
```
Comment (saberduck) [inline java-checks/src/main/java/org/sonar/java/checks/IterableIteratorCheck.java]: Message is empty?

**Pull request merged**